### DISCUSSION TITLE
Debug nodejs app when it runs inside Docker container

### DISCUSSION THREAD

**Discussion by GhostOrder28** [mfussenegger/nvim-dap#1201] [Q&A]
I managed to set nvim-dap to debug nodejs apps using [dapDebugServer.js](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#javascript), now I would like to go one step further and debug nodejs apps when they are running inside a docker container.

I've read the [guidelines](https://github.com/mfussenegger/nvim-dap/wiki/Local-and-Remote-Debugging-with-Docker) but still I can't make it work.

Just to compare this is the launch.json of my project which contain first the local debugging config (which works perfectly) and then a copy of the former with modifications I introduced according to what I was reading on internet about the subject.

Note: Port 4001 is the port I manually assigned to the node --inspect flag and it is exposed at the end of my Dockerfile

```
{
  "configurations": [
    {
      "name": "Attach: Local",
      "type": "pwa-node",
      "request": "attach",
      "skipFiles": ["<node_internals>/**"],
      "address": "localhost",
      "port": 4001,
      "cwd": "${workspaceFolder}",
      "restart": true
    },
    {
      "name": "Attach: Docker",
      "type": "pwa-node",
      "request": "attach",
      "mode": "remote",
      "skipFiles": ["<node_internals>/**"],
      "address": "localhost",
      "port": 4001,
      "cwd": "${workspaceFolder}",
      "preLaunchTask": "docker-run: debug",
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/app",
      "protocol": "inspector",
      "restart": true
    }
  ]
}
```

These are the relevant parts of the command I use to run the container.
```
docker run \
--rm \
-p 4000:4000 -p 4001:4001 \
--cap-add=SYS_PTRACE \
--security-opt seccomp=unconfined \
--network=mycustomnetwork \
app:tag
```

When I try to start a new debug session with the second configuration (the one for Docker) I get this error:

```
Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this)
or there is a problem with your adapter or `pwa-node` configuration.
Check the logs for errors (:help dap.set_log_level)
```

And when I look at the nodejs debugger adapter logs I can see:

```
Error: Could not connect to debug target at http://localhost:4001: Could not find any debuggable target
<stack trace here>
```

What I'm missing here?

**Comment by GhostOrder28** [comment]
So reading carefully the [guidelines](https://github.com/mfussenegger/nvim-dap/wiki/Local-and-Remote-Debugging-with-Docker) again it seems that the adapter needs to run inside the docker container, so I managed to install manually from the [official repository](https://github.com/microsoft/vscode-js-debug/tree/release/1.83), and now I can run it inside the container. However here is another hurdle, how can I tell nvim-dap that the debugger is running inside the container?

Currently this is my nvim-dap config for the adapter: 

```
dap.adapters["pwa-node"] = {
  type = "server",
  host = "::1",
  port = 8123,
}
```
As you can see I'm not using the executable key to run the adapter and that is because for some reason nvim-dap never calls the command I set so I'm my case I have to start the adapter manually in a separated terminal and it always the assigned port is always 8123, other than that all works just fine for debugging locally.

Now both the debugger (the one called with the node --inspect flag) and the adapter are running inside the docker container, all the relevant ports are exposed in the Dockerfile and mapped to localhost when I run the container: 

- 4000: for my application
- 4001: for the nodejs debugger (--inspect)
- 8123: for the adapter

But when I start a debugging session and choose the configuration for docker (the one I shared in the OP) I get immediately a ECONNRESET error and then:
```
Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this)
or there is a problem with your adapter or `pwa-node` configuration.
Check the logs for errors (:help dap.set_log_level)
```
My log level is ERROR and when I check the `~/.cache/nvim/dap.log` I only see empty lines. 

Any ideas of what can I do to tell nvim-dap to connect the adapter to the debugger when they are running inside the container?

**Comment by GhostOrder28** [comment]
I think I make a little bit of advance. I added the `preLaunchTask` property in my launch.json config with the label of a custom task I made to run the adapter inside a running docker container:

**launch.json**
```
{
  "name": "Attach: Docker",
  "type": "pwa-node",
  "request": "attach",
  "mode": "remote",
  "skipFiles": ["<node_internals>/**"],
  "address": "127.0.0.1",
  "port": 4001,
  "cwd": "${workspaceFolder}",
  "preLaunchTask": "Run adapter inside docker container",
  "localRoot": "${workspaceFolder}",
  "remoteRoot": "/app",
  "protocol": "inspector",
  "restart": true
}
```
**tasks.json**
```
{
  "tasks": [
    {
      "label": "Run adapter inside docker container",
      "type": "shell",
      "command": "docker exec -d <my container name> node /path/to/dapDebugServer.js"
    }
  ]
}
```
I have to run the command in detach mode otherwise nvim-dap gives me:
```
FAILURE Run adapter inside docker container
Failed to launch debugger; preLaunchTask 'Run adapter inside docker container' failed
```
Which is weird because if I inspect the process inside the container with the `top` command, I can see the process running:
```
PID  PPID USER     STAT   VSZ %VSZ CPU %CPU COMMAND
65     0 root     S    10.6g  43%   2   0% node /path/to/dapDebugServer.js
```
Also I can attach successfully to the process with `strace -p65 -s9999 -e write` but don't get any output, which leads me to think that although the adapter is running (because the process is running) it is not receiving any requests so far.

Also I don't get any output in the dap.log file.

This changes if I run the command in detach mode `-d`, then I no longer get an error from nvim-dap but I inmediately get the `Adapter disconnected` message, again if I check the container, the process is running and I can attach to it with the strace command.

However now when I inspect the dap.log file I can finally see a couple of logs:
```
[ DEBUG ] 2024-04-28T01:05:05Z-0500 ] ...nvim/site/pack/packer/start/nvim-dap/lua/dap/session.lua:1309 ]	"Connecting to debug adapter"	{
  host = "::1",
  port = 8123,
  type = "server"
}
[ DEBUG ] 2024-04-28T01:05:05Z-0500 ] ...nvim/site/pack/packer/start/nvim-dap/lua/dap/session.lua:1686 ]	"request"	{
  arguments = {
    adapterID = "nvim-dap",
    clientID = "neovim",
    clientName = "neovim",
    columnsStartAt1 = true,
    linesStartAt1 = true,
    locale = "en_US.UTF-8",
    pathFormat = "path",
    supportsProgressReporting = true,
    supportsRunInTerminalRequest = true,
    supportsStartDebuggingRequest = true,
    supportsVariableType = true
  },
  command = "initialize",
  seq = 1,
  type = "request"
}
```
So at least now nvim-dap is actually sending the message to the adapter to try to communicate with it, unfortunately there is no response from it.

Note: I'm 'hardcoding' the port to connect to the adapter just for now as every time I inspect the process with `netstat -anpe` I can see it always running on that port (that is also the case when I run the adapter locally), but of course ideally this should be dynamic.

Again this was a little advance but now I can double confirm that my issue is that nvim-dap cannot connect to the adapter, it is sending the message but there is no response. And well here tbh I don't have any clue about how to continue but at the same time I think that if I manage to solve this issue the remaining parts will be much easier.

**Comment by GhostOrder28** [comment]
I implemented process id selection, just to be sure that I was getting the right process, for this I had to find a way to get into the container processes which I did thanks to reading this [PR](https://github.com/mfussenegger/nvim-dap/pull/771), now when nvim-dap sends the 'initialize' message to the adapter it also sends the process id for the adapter in the message:
```
[ DEBUG ] 2024-04-28T01:05:05Z-0500 ] ...nvim/site/pack/packer/start/nvim-dap/lua/dap/session.lua:1309 ]	"Connecting to debug adapter"	{
  host = "::1",
  port = 8123,
  type = "server",
  pid = 65
}
```
However this didn't solve the issue, the adapter still is not responding, which led me to think that the problem could be actually the adapter (I know maybe I should've think on this at the beginning), so I tried a simple GET request to the adapter, this is the result when I do that inside the running container:
```
curl -v http://localhost:8123
* Host localhost:8123 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8123...
* Connected to localhost (::1) port 8123
> GET / HTTP/1.1
> Host: localhost:8123
> User-Agent: curl/8.5.0
> Accept: */*
>
```
And the server hangs, which means there is no actual response from the adapter, but the connection is being establish. Now if I make the request from localhost (outside the container) I get the **very same result** as above but with these extra lines:
```
* Request completely sent off
* Recv failure: Connection reset by peer
* Closing connection
curl: (56) Recv failure: Connection reset by peer
```
And just to be sure that the adapter is actually running I manually started the adapter in a separate terminal and it ran just fine with the usual message and no errors, then I repeat the curl requests both from inside the container and from outside, and get the same output that I just showed.
```
docker exec -it <my container name> /bin/sh
# node path/to/dapDebugServer.js
Debug server listening at ::1:8123
```
So to summarize, the connection to the server is being establish without errors but it is not responding.

**Reply by mfussenegger** [reply]
If you have neovim running outside the container, and dapDebugServer inside the container you need to make sure that the dapDebugServer binds itself to the public interface.
If it's binding to `::1` (loopback device), anything running outside the container won't be able to connect, unless you use docker with `--network host`.
